x/staking/src/keeper/redelegation.rs

```diff
@@ -0,0 +1,189 @@
+            if entry.is_mature(ctx_time) && !coin.amount.is_zero() {
+                balances.push(coin);
+            } else {
+                new_redelegations.push(entry);
+            }
+        }
+
+        // set the redelegation or remove it if there are no more entries
+        if new_redelegations.is_empty() {
+            self.remove_redelegation(ctx, &redelegation);
+        } else {
+            self.set_redelegation(ctx, &redelegation)?;
+        }
+        Ok(balances)
+    }
+
+    pub fn insert_redelegation_queue<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        redelegation: &Redelegation,
+        completion_time: chrono::DateTime<Utc>,
+    ) -> anyhow::Result<()> {
+        let mut time_slice = self.get_redelegation_queue_time_slice(ctx, completion_time)?;
+        let dvv_triplet = DvvTriplet {
+            del_addr: redelegation.delegator_address.clone(),
+            val_src_addr: redelegation.validator_src_address.clone(),
+            val_dst_addr: redelegation.validator_dst_address.clone(),
+        };
+        if time_slice.is_empty() {
+            self.set_redelegation_queue_time_slice(ctx, completion_time, vec![dvv_triplet])?;
+        } else {
+            time_slice.push(dvv_triplet);
+            self.set_redelegation_queue_time_slice(ctx, completion_time, time_slice)?;
+        }
+        Ok(())
+    }
+
+    pub fn get_redelegation_queue_time_slice<DB: Database, CTX: QueryableContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        completion_time: chrono::DateTime<Utc>,
+    ) -> anyhow::Result<Vec<DvvTriplet>> {
+        let store = ctx.kv_store(&self.store_key);
+        let store = store.prefix_store(REDELEGATION_QUEUE_KEY);
+
+        let key = completion_time
+            .timestamp_nanos_opt()
+            .expect("The timestamp_nanos_opt produces an integer that represents time in nanoseconds.
+                    The error in this method means that some system failure happened and the system cannot continue work.")
+            .to_ne_bytes();
+        if let Some(bytes) = store.get(&key) {
+            Ok(serde_json::from_slice(&bytes)?)
+        } else {
+            Ok(vec![])
+        }
+    }
+
+    pub fn set_redelegation_queue_time_slice<DB: Database, CTX: TransactionalContext<DB, SK>>(
+        &self,
+        ctx: &mut CTX,
+        completion_time: chrono::DateTime<Utc>,
+        redelegations: Vec<DvvTriplet>,
+    ) -> anyhow::Result<()> {
+        let store = ctx.kv_store_mut(&self.store_key);
+        let mut store = store.prefix_store_mut(REDELEGATION_QUEUE_KEY);
+
+        let key = completion_time
+            .timestamp_nanos_opt()
+            .expect("The timestamp_nanos_opt produces an integer that represents time in nanoseconds.
+                    The error in this method means that some system failure happened and the system cannot continue work.")
```

Comment:
The expect message should be something like "provided the date is between 1677-09-21T00:12:43.145224192 and 2262-04-11T23:47:16.854775807 the conversion won't fail"

Reply:
I'm starting to doubt whether panicking is the right thing to do here. It could be a vulnerability that allows attackers to halt a chain. Where does the `completion_time` come from? Perhaps we should parse those times upfront to guarantee that they can be converted to an i64?